zkevm-circuits/src/evm_circuit/util/constraint_builder.rs

```diff
@@ -0,0 +1,559 @@
+    }
+
+    pub(crate) fn opcode_lookup_at(
+        &mut self,
+        index: Expression<F>,
+        opcode: Expression<F>,
+        is_code: Expression<F>,
+    ) {
+        let is_root_create =
+            self.curr.state.is_root.expr() * self.curr.state.is_create.expr();
+        self.add_constraint(
+            "The opcode source when is_root and is_create (Root creation transaction) is not determined yet",
+            is_root_create.clone(),
+        );
+        self.add_lookup(
+            Lookup::Bytecode {
+                hash: self.curr.state.opcode_source.expr(),
+                index,
+                value: opcode,
+                is_code,
+            }
+            .conditional(1.expr() - is_root_create),
+        );
+    }
+
+    // Rw
+
+    fn rw_lookup_inner(
+        &mut self,
+        counter: Expression<F>,
+        is_write: Expression<F>,
+        tag: Expression<F>,
+        values: [Expression<F>; 5],
+    ) {
+        self.add_lookup(Lookup::Rw {
+            counter,
+            is_write,
+            tag,
+            values,
+        });
+    }
+
+    fn rw_lookup(
+        &mut self,
+        is_write: Expression<F>,
+        tag: Expression<F>,
+        values: [Expression<F>; 5],
+    ) {
+        self.rw_lookup_inner(
+            self.curr.state.rw_counter.expr() + self.rw_counter_offset.expr(),
+            is_write,
+            tag,
+            values,
+        );
+        self.rw_counter_offset += 1;
+    }
+
+    // Stack
+
+    pub(crate) fn stack_pop(&mut self, value: Expression<F>) {
+        self.stack_lookup(
+            false.expr(),
+            self.stack_pointer_offset.expr(),
+            value,
+        );
+        self.stack_pointer_offset += 1;
+    }
+
+    pub(crate) fn stack_push(&mut self, value: Expression<F>) {
+        self.stack_pointer_offset -= 1;
+        self.stack_lookup(true.expr(), self.stack_pointer_offset.expr(), value);
+    }
+
+    pub(crate) fn stack_lookup(
+        &mut self,
+        is_write: Expression<F>,
+        stack_pointer_offset: Expression<F>,
+        value: Expression<F>,
+    ) {
+        self.rw_lookup(
+            is_write,
+            RwTableTag::Stack.expr(),
+            [
+                self.curr.state.call_id.expr(),
+                self.curr.state.stack_pointer.expr() + stack_pointer_offset,
+                value,
+                0.expr(),
+                0.expr(),
+            ],
+        );
+    }
+
+    // Memory
+
+    pub(crate) fn memory_lookup(
+        &mut self,
+        is_write: Expression<F>,
+        memory_address: Expression<F>,
+        byte: Expression<F>,
+    ) {
+        self.rw_lookup(
+            is_write,
+            RwTableTag::Memory.expr(),
+            [
+                self.curr.state.call_id.expr(),
+                memory_address,
+                byte,
+                0.expr(),
+                0.expr(),
+            ],
+        );
+    }
+
+    pub(crate) fn memory_lookup_inner(
```

Comment:
I found the func `memory_lookup_inner` can be confusing. It has quite subtle difference from `memory_lookup` as it doesn't increase the `rw_counter_offset`. So are the `rw_lookup` and `rw_lookup_inner`. I wonder if you could come up with a better design or name for this? Exposing inner function to public is also a bit weird. Probably you can rename it to `memory_lookup_with_counter` and add doc with something like "the func doesn't increase rw_counter_offset".

Reply:
Previously it's `memory_lookup_at`, I forgot why I changed it to `inner`...   Let me update it to `memory_lookup_with_counter` with some comments, thanks!